src/vs/workbench/contrib/debug/browser/debugCommands.ts

```diff
@@ -179,6 +187,92 @@ async function changeDebugConsoleFocus(accessor: ServicesAccessor, next: boolean
 	}
 }
 
+async function navigateCallStack(debugService: IDebugService, down: boolean) {
+	const frame = debugService.getViewModel().focusedStackFrame;
+	if (frame) {
+
+		let callStack = frame.thread.getCallStack();
+		const index = callStack.findIndex(elem => elem.frameId === frame.frameId);
+		let nextVisibleFrame;
+		if (down) {
+			if (index >= callStack.length - 1) {
+				if ((<Thread>frame.thread).reachedEndOfCallStack) {
+					goToTopOfCallStack(debugService);
+					return;
+				} else {
+					await (<DebugModel>debugService.getModel()).fetchAndRefreshCallStack((<Thread>frame.thread), 20);
+					callStack = frame.thread.getCallStack();
+				}
+			}
+			nextVisibleFrame = findNextVisibleFrame(true, callStack, index + 1);
+		} else {
+			if (index <= 0) {
+				goToBottomOfCallStack(debugService);
+				return;
+			}
+			// goes to bottom of visible panes to reduce performance overhead
+			nextVisibleFrame = findNextVisibleFrame(false, callStack, index - 1);
+		}
+
+		if (nextVisibleFrame) {
+			debugService.focusStackFrame(nextVisibleFrame);
+		}
+	}
+}
+
+async function goToBottomOfCallStack(debugService: IDebugService) {
+	const thread = debugService.getViewModel().focusedThread;
+	if (thread) {
+		await (<DebugModel>debugService.getModel()).fetchAndRefreshCallStack((<Thread>thread));
+		const callStack = thread.getCallStack();
+		if (callStack.length > 0) {
+			const nextVisibleFrame = findNextVisibleFrame(false, callStack, callStack.length - 1);
+			if (nextVisibleFrame) {
+				debugService.focusStackFrame(nextVisibleFrame);
+			}
+		}
+	}
+}
+
+function goToTopOfCallStack(debugService: IDebugService) {
+	const thread = debugService.getViewModel().focusedThread;
+
+	if (thread) {
+		debugService.focusStackFrame(thread.getTopStackFrame());
+	}
+}
+
+/**
+ * Finds next frame that is not skipped by SkipFiles
+ * @param down specifies whether to search downwards if the current file is skipped.
+ * @param callStack the call stack to search
+ * @param initialIndex the index to start the search at
+ */
+function findNextVisibleFrame(down: boolean, callStack: readonly IStackFrame[], initialIndex: number) {
+	let index = initialIndex;
+	let currFrame;
+	do {
+		currFrame = callStack[index];
+		if (!(currFrame.source.presentationHint === 'deemphasize' || currFrame.presentationHint === 'deemphasize')) {
```

Comment:
couldn't actually call the isDeemphasized function in callStackView because exporting the function and using it here would cause a circular dependency